graspologic/partition/leiden.py

```diff
@@ -0,0 +1,402 @@
+    if is_weighted is not None and not isinstance(is_weighted, bool):
+        raise TypeError("is_weighted must either be a bool or None")
+    if not isinstance(weight_default, (int, float)):
+        raise TypeError("weight_default must be a float")
+    if not isinstance(check_directed, bool):
+        raise TypeError("check_directed must be a bool")
+
+    if iterations < 1:
+        raise ValueError("iterations must be a positive integer")
+    if resolution <= 0:
+        raise ValueError("resolution must be a positive float")
+    if randomness <= 0:
+        raise ValueError("randomness must be a positive float")
+    if random_seed is not None and random_seed <= 0:
+        raise ValueError(
+            "random_seed must be a positive integer (the native PRNG implementation is an unsigned 64 bit integer)"
+        )
+
+
+def leiden(
+    graph: Union[
+        List[Tuple[Any, Any, Union[int, float]]],
+        nx.Graph,
+        np.ndarray,
+        scipy.sparse.csr.csr_matrix,
+    ],
+    starting_communities: Optional[Dict[str, int]] = None,
+    iterations: int = 1,
+    resolution: float = 1.0,
+    randomness: float = 0.001,
+    use_modularity: bool = True,
+    random_seed: Optional[int] = None,
+    weight_attribute: str = "weight",
+    is_weighted: Optional[bool] = None,
+    weight_default: float = 1.0,
+    check_directed: bool = True,
+) -> Dict[str, int]:
+    """
+    Leiden is a global network partitioning algorithm. Given a list of edges and a maximization function, it will
+    iterate through the network attempting to find an optimal partitioning of the entire network.
+
+    Parameters
+    ----------
+    graph : Union[List[Tuple[Any, Any, Union[int, float]]], nx.Graph, np.ndarray, scipy.sparse.csr.csr_matrix]
+        A graph representation, whether a weighted edge list referencing an undirected graph, an undirected networkx
+        graph, or an undirected adjacency matrix in either numpy.ndarray or scipy.sparse.csr.csr_matrix form.
+    starting_communities : Optional[Dict[str, int]]
+        Default is ``None``. An optional community mapping dictionary that contains the string representation of the
+        node and the community it belongs to. Note this map must contain precisely the same nodes as the graph and every
+        node must have a positive community id. This mapping forms the starting point of Leiden clustering and can be
+        very useful in saving the state of a previous partition schema from a previous graph and then adjusting the
+        graph based on new temporal data (additions, removals, weight changes, connectivity changes, etc).
+        New nodes should get their own unique community positive integer, but the original partition can be very useful
+        to speed up future runs of leiden. If no community map is provided, the default behavior is to create a
+        node community identity map, where every node is in their own community.
+    iterations : int
```

Comment:
there's a case for calling this `n_init` like in sklearn, iterations usually refers to something else for most algos. Although the fact that this restarts from the last partition makes this stickier. In sklearn, this is called `warm_start`.

One bigger question - when you say the "last run" is that literally what you mean? If we number each run by order it was ran in, is it true that using this scheme `... Q_3 > Q_2 > Q_1` where Q is modularity? So that's why it always makes sense to start with the last run, as opposed to the best run, since best = last?

Reply:
I'm not sure `n_init` is the right concept though.  Basically in this algorithm, it runs through the entire process until the quality score is maximized and the next iteration doesn't result in any movement.  That whole thing counts as 1 iteration, even though you looped potentially thousands of times.

Setting it > 1 just says "okay, I know you did this 1000 times, but I want you to try iterations - 1 more times for me, ok?"

I was just sticking with the nomenclature from the reference implementation @ https://github.com/CWTSLeiden/networkanalysis/blob/master/src/main/java/nl/cwts/networkanalysis/run/RunNetworkClustering.java#L137

Reply:
I'm fine with changing it but `n_init` seems even less ideal than `iterations`, which is also not really a great choice.  Maybe... `extra_iterations`?  I'd have to change the logic a bit and make it an optional parameter, but that seems to make more sense to me